funding/commitment_type_negotiation.go

```diff
@@ -0,0 +1,119 @@
+package funding
+
+import (
+	"errors"
+
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// errUnsupportedExplicitNegotiation is an error returned when explicit
+	// channel commitment negotiation is attempted but either peer of the
+	// channel does not support it.
+	errUnsupportedExplicitNegotiation = errors.New("explicit channel " +
+		"type negotiation not supported")
+
+	// errUnsupportedCommitmentType is an error returned when a specific
+	// channel commitment type is being explicitly negotiated but either
+	// peer of the channel does not support it.
+	errUnsupportedChannelType = errors.New("requested channel type " +
+		"not supported")
+)
+
+// negotiateCommitmentType negotiates the commitment type of a newly opened
+// channel. If a channelType is provided, explicit negotiation for said type
+// will be attempted if the set of both local and remote features support it.
+// Otherwise, implicit negotiation will be attempted.
+func negotiateCommitmentType(channelType *lnwire.ChannelType,
+	local, remote *lnwire.FeatureVector) (lnwallet.CommitmentType, error) {
+
+	if channelType != nil {
+		if !hasFeatures(local, remote, lnwire.ExplicitChannelTypeOptional) {
+			return 0, errUnsupportedExplicitNegotiation
+		}
+		return explicitNegotiateCommitmentType(
+			*channelType, local, remote,
+		)
+	}
+
+	return implicitNegotiateCommitmentType(local, remote), nil
+}
+
+// explicitNegotiateCommitmentType attempts to explicitly negotiate for a
+// specific channel type. Since the channel type is comprised of a set of even
+// feature bits, we also make sure each feature is supported by both peers. An
+// error is returned if either peer does not support said channel type.
+func explicitNegotiateCommitmentType(channelType lnwire.ChannelType,
+	local, remote *lnwire.FeatureVector) (lnwallet.CommitmentType, error) {
+
+	// No features, use legacy commitment type.
+	channelFeatures := lnwire.RawFeatureVector(channelType)
+	if channelFeatures.OnlyContains() {
```

Comment:
Won't this always return false? No bits passed in as arg, so the length check fails.

Reply:
I think maybe we should go with a different name for this method also, not that clear IMO what it does at a glance w/o needing refer back to the implementation details.

Reply:
Perhaps we should put certain behavior behind a build flag also? Can't think of a scenario outside of the itests that we actually want to allow using the legacy channel type.

Reply:
> Won't this always return false? No bits passed in as arg, so the length check fails.

The length check wouldn't fail since no bits are provided as arguments and a legacy channel has no features.

> I think maybe we should go with a different name for this method also, not that clear IMO what it does at a glance w/o needing refer back to the implementation details.

I agree, renamed to `IsEmpty`.

> Perhaps we should put certain behavior behind a build flag also? Can't think of a scenario outside of the itests that we actually want to allow using the legacy channel type.

Sounds good to me. Without explicit negotiation though, we already allow legacy channels to be opened if we downgrade the `static_remote_key` feature from required to optional to a peer we have an already opened legacy channel with. It seems this build flag would also need to handle that?